README.md
# parse-yapp
Parse::Yapp, Yet Another Parser Parser for Perl


Changes
Revision history for Perl extension Yapp.

	- original version; created by h2xs 1.18
0.01  Thu Jun 25 20:02:09 1998
0.02  Never released
0.03  Never released
0.04  Never released
	- Fix installation of yapp.pl to $INST_SCRIPT
	- $VERSION is now in Yapp/Driver to check compatibility
	- Add debugging driver and debug option
	- Bug in error recovery fixed: do not shift an $error reduction
	- Add LeftValue, Curtok and Curval methods
	- Add driver version compatibility check 
0.05  Thu Jul 03 20:05:05 1998
	- Add LeftValue, Curtok and Curval methods
0.06  Tue Jul 07 20:36:17 GMT 1998
    - Error token '$error' becomes 'error' (like in yacc)
    - The '$end' token becomes ''
0.07  Never released
    - Default action become a separate entry in states hash
    - $COMPATIBLE value changed to 0.08
    - The grammar parser is now a Yapp parser module
    - Comments can be either Perl (#...\n) or C /* ... */ style
    - The parser accepts %type, %union and <type> constructs and
      almost ignore them, except checking consistency between token
      and non-terminal declarations (warnings)
    - The parser now has error recovery and consistent error line numbers
    - The parser now accepts "in rule" actions and generates pseudo
      empty rules to reduce, named @x-y, where x is a sequential
      number and y the 'dot position' in the rule for the driver
      to know how many parameters to pass to semantic action sub
    - Add "in rule" actions handling in Driver.pm
    - Empty rhs need not be the first one anymore
    - Warning if more than one empty rhs for a given lhs
0.08  Fri Jul 10 22:04:31 GMT 1998
    - Changed 'print STDERR' to 'warn' in parser.
    - Use of literal 'error' produces a warning and is treated as
      the error token
    - Add prefix 'YY' before each parser object methods to avoid
      clashes with user written methods
    - Renamed YYUserData to YYData (shorter and more consistent with
      other methods names)
    - Renamed YYLeftVal to YYSemval for same reasons
    - Modified Driver.pm so Semval(-n..n) reacts like $-n..$n in yacc
0.09  Never released
    - Changed test suite to 't/' style and add base tests for semantic
      actions/values and associativity tests
    - Check code to be (almost) Perl's -w compatible
    - Updates to pod section in Yapp.pm reflecting most of those changes
0.10  Mon Jul 13 20:53:40 GMT 1998
    - Cosmetic changes
0.11  Wed Jul 15 19:46:17 GMT 1998
    - Renamed Yapp::Parse parameters with a leading yy
    - Updated Yapp.pm pod section
0.12  Tue Jul 21 22:34:00 GMT 1998
0.13  Never released. (I'm not supersticious, but who knows...8-))
    - Renaming to Parse::Yapp for better CPAN entry...
0.14  Wed Jul 29 21:43:03 GMT 1998
    - Doc change: empty token is '', not undef (perl -w complains otherwise)
    - Bug in _ReduceGrammar: used $ruleno instead of $#{$grammar{RULES}}
      making no-terminal pointing to wrong rhs if useless rules.
0.15  Mon Aug 17 11:39:01 CEST 1998
    - YappParse.yp modified to allow empty tail section (not even \n)
    - YappParse.yp modified to diagnose lack of rules in grammar section
    - Driver.pm has been modified so there is no performance impact at

lib/Parse/Yapp/Parse.pm
####################################################################
#
#    This file was generated using Parse::Yapp version 1.05.
#
#        Don't edit this file, use source file instead.
#
#             ANY CHANGE MADE HERE WILL BE LOST !
#
####################################################################
package Parse::Yapp::Parse;
use vars qw ( @ISA );
use strict;

@ISA= qw ( Parse::Yapp::Driver );
use Parse::Yapp::Driver;

#line 1 "YappParse.yp"

# Copyright © 1998, 1999, 2000, 2001, Francois Desarmenien.
# Copyright © 2017 William N. Braswell, Jr.
# All Rights Reserved.
# (see COPYRIGHT in Parse::Yapp.pm pod section for use and distribution rights)
#
# Parse/Yapp/Parser.yp: Parse::Yapp::Parser.pm source file
#
# Use: yapp -m 'Parse::Yapp::Parse' -o Parse/Yapp/Parse.pm YappParse.yp
#
# to generate the Parser module.
# 
#line 15 "YappParse.yp"

require 5.004;

use Carp;

my($input,$lexlevel,@lineno,$nberr,$prec,$labelno);
my($syms,$head,$tail,$token,$term,$nterm,$rules,$precterm,$start,$nullable);
my($expect);



sub new {
        my($class)=shift;
        ref($class)
    and $class=ref($class);

    my($self)=$class->SUPER::new( yyversion => '1.05',
                                  yystates =>
[
	{#State 0
		ACTIONS => {
			'error' => 12,
			'START' => 10,
			"\n" => 9,
			'TYPE' => 8,
			'TOKEN' => 13,
			"%%" => -6,
			'EXPECT' => 2,
			'HEADCODE' => 3,
			'ASSOC' => 4,
			'UNION' => 6
		},
		GOTOS => {
			'head' => 5,
			'decl' => 14,
			'yapp' => 1,
			'headsec' => 7,
			'decls' => 11
		}
	},
	{#State 1
		ACTIONS => {
			'' => 15
		}
	},
	{#State 2
		ACTIONS => {
			'NUMBER' => 16
		}
	},
	{#State 3
		ACTIONS => {
			"\n" => 17
		}
	},
	{#State 4
		ACTIONS => {
			"<" => 18
		},
		DEFAULT => -19,
		GOTOS => {
			'typedecl' => 19
		}
	},
	{#State 5
		ACTIONS => {
			'error' => 22,
			'IDENT' => 25,
			"%%" => 24
		},
		GOTOS => {
			'rules' => 21,
			'body' => 23,
			'rulesec' => 20
		}
	},
	{#State 6
		ACTIONS => {
			'CODE' => 26
		}
	},
	{#State 7
		ACTIONS => {
			"%%" => 27
		}
	},
	{#State 8
		ACTIONS => {
			"<" => 18
		},
		DEFAULT => -19,
		GOTOS => {
			'typedecl' => 28
		}
	},
	{#State 9
		DEFAULT => -10
	},
	{#State 10
		ACTIONS => {
			'IDENT' => 29
		},
		GOTOS => {
			'ident' => 30
		}
	},
	{#State 11
		ACTIONS => {
			'EXPECT' => 2,
			'HEADCODE' => 3,
			'ASSOC' => 4,
			'UNION' => 6,
			"\n" => 9,
			'TYPE' => 8,
			'START' => 10,
			'error' => 12,
			'TOKEN' => 13,
			"%%" => -7
		},
		GOTOS => {
			'decl' => 31
		}
	},
	{#State 12
		ACTIONS => {
			"\n" => 32
		}
	},
	{#State 13
		ACTIONS => {
			"<" => 18
		},
		DEFAULT => -19,
		GOTOS => {
			'typedecl' => 33
		}
	},
	{#State 14
		DEFAULT => -9
	},
	{#State 15
		DEFAULT => 0
	},
	{#State 16
		ACTIONS => {
			"\n" => 34
		}
	},
	{#State 17
		DEFAULT => -14
	},
	{#State 18
		ACTIONS => {
			'IDENT' => 35
		}
	},
	{#State 19
		ACTIONS => {
			'LITERAL' => 36,
			'IDENT' => 29
		},
		GOTOS => {
			'symbol' => 39,
			'ident' => 38,
			'symlist' => 37
		}
	},
	{#State 20
		ACTIONS => {
			"%%" => 40,
			'IDENT' => 25,
			'error' => 22
		},
		GOTOS => {
			'rules' => 41
		}
	},
	{#State 21
		DEFAULT => -28
	},
	{#State 22
		ACTIONS => {
			";" => 42
		}
	},
	{#State 23
		ACTIONS => {
			'TAILCODE' => 43
		},
		DEFAULT => -45,
		GOTOS => {
			'tail' => 44
		}
	},
	{#State 24
		DEFAULT => -26
	},
	{#State 25
		ACTIONS => {
			":" => 45
		}
	},
	{#State 26
		ACTIONS => {
			"\n" => 46
		}
	},
	{#State 27
		DEFAULT => -5
	},
	{#State 28
		ACTIONS => {
			'IDENT' => 29
		},
		GOTOS => {
			'ident' => 48,
			'identlist' => 47
		}
	},
	{#State 29
		DEFAULT => -4
	},
	{#State 30
		ACTIONS => {
			"\n" => 49
		}
	},
	{#State 31
		DEFAULT => -8
	},
	{#State 32
		DEFAULT => -18
	},
	{#State 33
		ACTIONS => {
			'LITERAL' => 36,
			'IDENT' => 29
		},
		GOTOS => {
			'ident' => 38,
			'symlist' => 50,
			'symbol' => 39
		}
	},
	{#State 34
		DEFAULT => -17
	},
	{#State 35
		ACTIONS => {
			">" => 51
		}
	},
	{#State 36
		DEFAULT => -2
	},
	{#State 37
		ACTIONS => {
			'LITERAL' => 36,
			"\n" => 53,
			'IDENT' => 29
		},
		GOTOS => {
			'symbol' => 52,
			'ident' => 38
		}
	},
	{#State 38
		DEFAULT => -3
	},
	{#State 39
		DEFAULT => -22
	},
	{#State 40
		DEFAULT => -25
	},
	{#State 41
		DEFAULT => -27
	},
	{#State 42
		DEFAULT => -30
	},
	{#State 43
		DEFAULT => -46
	},
	{#State 44
		DEFAULT => -1
	},
	{#State 45
		ACTIONS => {
			'CODE' => 61,
			'LITERAL' => 36,
			'IDENT' => 29
		},
		DEFAULT => -35,
		GOTOS => {
			'ident' => 38,
			'rhss' => 59,
			'code' => 60,
			'symbol' => 54,
			'rule' => 55,
			'rhselt' => 56,
			'rhs' => 57,
			'rhselts' => 58
		}
	},
	{#State 46
		DEFAULT => -15
	},
	{#State 47
		ACTIONS => {
			"\n" => 62,
			'IDENT' => 29
		},
		GOTOS => {
			'ident' => 63
		}
	},
	{#State 48
		DEFAULT => -24
	},
	{#State 49
		DEFAULT => -13
	},
	{#State 50
		ACTIONS => {
			'IDENT' => 29,
			"\n" => 64,
			'LITERAL' => 36
		},
		GOTOS => {
			'symbol' => 52,
			'ident' => 38
		}
	},
	{#State 51
		DEFAULT => -20
	},
	{#State 52
		DEFAULT => -21
	},
	{#State 53
		DEFAULT => -12
	},
	{#State 54
		DEFAULT => -39
	},
	{#State 55
		DEFAULT => -32
	},
	{#State 56
		DEFAULT => -38
	},
	{#State 57
		ACTIONS => {
			'PREC' => 65
		},
		DEFAULT => -34,
		GOTOS => {
			'prec' => 66
		}
	},
	{#State 58
		ACTIONS => {
			'CODE' => 61,
			'LITERAL' => 36,
			'IDENT' => 29
		},
		DEFAULT => -36,
		GOTOS => {
			'code' => 60,
			'rhselt' => 67,
			'ident' => 38,
			'symbol' => 54
		}
	},
	{#State 59
		ACTIONS => {
			";" => 68,
			"|" => 69
		}
	},
	{#State 60
		DEFAULT => -40
	},
	{#State 61
		DEFAULT => -44
	},
	{#State 62
		DEFAULT => -16
	},
	{#State 63
		DEFAULT => -23
	},
	{#State 64
		DEFAULT => -11
	},
	{#State 65
		ACTIONS => {
			'IDENT' => 29,
			'LITERAL' => 36
		},
		GOTOS => {
			'ident' => 38,
			'symbol' => 70
		}
	},
	{#State 66
		ACTIONS => {
			'CODE' => 61
		},
		DEFAULT => -42,
		GOTOS => {
			'code' => 71,
			'epscode' => 72
		}
	},
	{#State 67
		DEFAULT => -37
	},
	{#State 68
		DEFAULT => -29
	},
	{#State 69
		ACTIONS => {
			'IDENT' => 29,
			'LITERAL' => 36,
			'CODE' => 61
		},
		DEFAULT => -35,
		GOTOS => {
			'rhs' => 57,
			'code' => 60,
			'rhselts' => 58,
			'symbol' => 54,
			'rule' => 73,
			'rhselt' => 56,
			'ident' => 38
		}
	},
	{#State 70
		DEFAULT => -41
	},
	{#State 71
		DEFAULT => -43
	},
	{#State 72
		DEFAULT => -33
	},
	{#State 73
		DEFAULT => -31
	}
],
                                  yyrules  =>
[
	[#Rule 0
		 '$start', 2, undef
	],
	[#Rule 1
		 'yapp', 3, undef
	],
	[#Rule 2
		 'symbol', 1,
sub
#line 33 "YappParse.yp"
{
                        exists($$syms{$_[1][0]})
                    or  do {
                        $$syms{$_[1][0]} = $_[1][1];
                        $$term{$_[1][0]} = undef;
                    };
                    $_[1]
                }
	],
	[#Rule 3
		 'symbol', 1, undef
	],
	[#Rule 4
		 'ident', 1,
sub
#line 44 "YappParse.yp"
{
                        exists($$syms{$_[1][0]})
                    or  do {
                        $$syms{$_[1][0]} = $_[1][1];
                        $$term{$_[1][0]} = undef;
                    };
                    $_[1]
                }
	],
	[#Rule 5
		 'head', 2, undef
	],
	[#Rule 6
		 'headsec', 0, undef
	],
	[#Rule 7
		 'headsec', 1, undef
	],
	[#Rule 8
		 'decls', 2, undef
	],
	[#Rule 9
		 'decls', 1, undef
	],
	[#Rule 10
		 'decl', 1, undef
	],
	[#Rule 11
		 'decl', 4,
sub
#line 69 "YappParse.yp"
{
                for (@{$_[3]}) {
                    my($symbol,$lineno)=@$_;

                        exists($$token{$symbol})
                    and do {
                        _SyntaxError(0,
                                "Token $symbol redefined: ".
                                "Previously defined line $$syms{$symbol}",
                                $lineno);
                        next;
                    };
                    $$token{$symbol}=$lineno;
                    $$term{$symbol} = [ ];
                }
                undef
            }
	],
	[#Rule 12
		 'decl', 4,
sub
#line 87 "YappParse.yp"
{
                for (@{$_[3]}) {
                    my($symbol,$lineno)=@$_;

                        defined($$term{$symbol}[0])
                    and do {
                        _SyntaxError(1,
                            "Precedence for symbol $symbol redefined: ".
                            "Previously defined line $$syms{$symbol}",
                            $lineno);
                        next;
                    };
                    $$token{$symbol}=$lineno;
                    $$term{$symbol} = [ $_[1][0], $prec ];
                }
                ++$prec;
                undef
            }
	],
	[#Rule 13
		 'decl', 3,
sub
#line 105 "YappParse.yp"
{ $start=$_[2][0]; undef }
	],
	[#Rule 14
		 'decl', 2,
sub
#line 106 "YappParse.yp"
{ push(@$head,$_[1]); undef }
	],
	[#Rule 15
		 'decl', 3,
sub
#line 107 "YappParse.yp"
{ undef }
	],
	[#Rule 16
		 'decl', 4,
sub
#line 109 "YappParse.yp"
{
                for ( @{$_[3]} ) {
                    my($symbol,$lineno)=@$_;

                        exists($$nterm{$symbol})
                    and do {
                        _SyntaxError(0,
                                "Non-terminal $symbol redefined: ".
                                "Previously defined line $$syms{$symbol}",
                                $lineno);
                        next;
                    };
                    delete($$term{$symbol});   #not a terminal
                    $$nterm{$symbol}=undef;    #is a non-terminal
                }
            }
	],
	[#Rule 17
		 'decl', 3,
sub
#line 125 "YappParse.yp"
{ $expect=$_[2][0]; undef }
	],
	[#Rule 18
		 'decl', 2,
sub
#line 126 "YappParse.yp"
{ $_[0]->YYErrok }
	],
	[#Rule 19
		 'typedecl', 0, undef
	],
	[#Rule 20
		 'typedecl', 3, undef
	],
	[#Rule 21
		 'symlist', 2,
sub
#line 133 "YappParse.yp"
{ push(@{$_[1]},$_[2]); $_[1] }
	],
	[#Rule 22
		 'symlist', 1,
sub
#line 134 "YappParse.yp"
{ [ $_[1] ] }
	],
	[#Rule 23
		 'identlist', 2,
sub
#line 137 "YappParse.yp"
{ push(@{$_[1]},$_[2]); $_[1] }
	],
	[#Rule 24
		 'identlist', 1,
sub
#line 138 "YappParse.yp"
{ [ $_[1] ] }
	],
	[#Rule 25
		 'body', 2,
sub
#line 143 "YappParse.yp"
{
                    $start
                or  $start=$$rules[1][0];

                    ref($$nterm{$start})
                or  _SyntaxError(2,"Start symbol $start not found ".
                                   "in rules section",$_[2][1]);

                $$rules[0]=[ '$start', [ $start, chr(0) ], undef, undef ];
            }
	],
	[#Rule 26
		 'body', 1,
sub
#line 153 "YappParse.yp"
{ _SyntaxError(2,"No rules in input grammar",$_[1][1]); }
	],
	[#Rule 27
		 'rulesec', 2, undef
	],
	[#Rule 28
		 'rulesec', 1, undef
	],
	[#Rule 29
		 'rules', 4,
sub
#line 160 "YappParse.yp"
{ _AddRules($_[1],$_[3]); undef }
	],
	[#Rule 30
		 'rules', 2,
sub
#line 161 "YappParse.yp"
{ $_[0]->YYErrok }
	],
	[#Rule 31
		 'rhss', 3,
sub
#line 164 "YappParse.yp"
{ push(@{$_[1]},$_[3]); $_[1] }
	],
	[#Rule 32
		 'rhss', 1,
sub
#line 165 "YappParse.yp"
{ [ $_[1] ] }
	],
	[#Rule 33
		 'rule', 3,
sub
#line 168 "YappParse.yp"
{ push(@{$_[1]}, $_[2], $_[3]); $_[1] }
	],
	[#Rule 34
		 'rule', 1,
sub
#line 169 "YappParse.yp"
{
                                my($code)=undef;

                                    defined($_[1])
                                and $_[1][-1][0] eq 'CODE'
                                and $code = ${pop(@{$_[1]})}[1];

                                push(@{$_[1]}, undef, $code);

                                $_[1]
                            }
	],
	[#Rule 35
		 'rhs', 0, undef
	],
	[#Rule 36
		 'rhs', 1, undef
	],
	[#Rule 37
		 'rhselts', 2,
sub
#line 186 "YappParse.yp"
{ push(@{$_[1]},$_[2]); $_[1] }
	],
	[#Rule 38
		 'rhselts', 1,
sub
#line 187 "YappParse.yp"
{ [ $_[1] ] }
	],
	[#Rule 39
		 'rhselt', 1,
sub
#line 190 "YappParse.yp"
{ [ 'SYMB', $_[1] ] }
	],
	[#Rule 40
		 'rhselt', 1,
sub
#line 191 "YappParse.yp"
{ [ 'CODE', $_[1] ] }
	],
	[#Rule 41
		 'prec', 2,
sub
#line 195 "YappParse.yp"
{
                       	defined($$term{$_[2][0]})
                    or  do {
                        _SyntaxError(1,"No precedence for symbol $_[2][0]",
                                         $_[2][1]);
                        return undef;
                    };

                    ++$$precterm{$_[2][0]};
                    $$term{$_[2][0]}[1];
				}
	],
	[#Rule 42
		 'epscode', 0,
sub
#line 208 "YappParse.yp"
{ undef }
	],
	[#Rule 43
		 'epscode', 1,
sub
#line 209 "YappParse.yp"
{ $_[1] }
	],
	[#Rule 44
		 'code', 1,
sub
#line 212 "YappParse.yp"
{ $_[1] }
	],
	[#Rule 45
		 'tail', 0, undef
	],
	[#Rule 46
		 'tail', 1,
sub
#line 218 "YappParse.yp"
{ $tail=$_[1] }
	]
],
                                  @_);
    bless($self,$class);
}

#line 221 "YappParse.yp"

sub _Error {
    my($value)=$_[0]->YYCurval;

    my($what)= $token ? "input: '$$value[0]'" : "end of input";

    _SyntaxError(1,"Unexpected $what",$$value[1]);
}

sub _Lexer {
 
    #At EOF
        pos($$input) >= length($$input)
    and return('',[ undef, -1 ]);

    #In TAIL section
        $lexlevel > 1
    and do {
        my($pos)=pos($$input);

        $lineno[0]=$lineno[1];
        $lineno[1]=-1;
        pos($$input)=length($$input);
        return('TAILCODE',[ substr($$input,$pos), $lineno[0] ]);
    };

    #Skip blanks
            $lexlevel == 0
        ?   $$input=~m{\G((?:
                                [\t\ ]+    # Any white space char but \n
                            |   \#[^\n]*  # Perl like comments
                            |   /\*.*?\*/ # C like comments
                            )+)}xsgc
        :   $$input=~m{\G((?:
                                \s+       # any white space char
                            |   \#[^\n]*  # Perl like comments
                            |   /\*.*?\*/ # C like comments
                            )+)}xsgc
    and do {
        my($blanks)=$1;

        #Maybe At EOF
            pos($$input) >= length($$input)
        and return('',[ undef, -1 ]);

        $lineno[1]+= $blanks=~tr/\n//;
    };

    $lineno[0]=$lineno[1];

        $$input=~/\G([A-Za-z_][A-Za-z0-9_]*)/gc
    and return('IDENT',[ $1, $lineno[0] ]);

        $$input=~/\G('(?:[^'\\]|\\\\|\\'|\\)+?')/gc
    and do {
            $1 eq "'error'"
        and do {
            _SyntaxError(0,"Literal 'error' ".
                           "will be treated as error token",$lineno[0]);
            return('IDENT',[ 'error', $lineno[0] ]);
        };
        return('LITERAL',[ $1, $lineno[0] ]);
    };

        $$input=~/\G(%%)/gc
    and do {
        ++$lexlevel;
        return($1, [ $1, $lineno[0] ]);
    };

        $$input=~/\G\{/gc
    and do {
        my($level,$from,$code);

        $from=pos($$input);

        $level=1;
        while($$input=~/([{}])/gc) {
                substr($$input,pos($$input)-1,1) eq '\\' #Quoted
            and next;
                $level += ($1 eq '{' ? 1 : -1)
            or last;
        }
            $level
        and  _SyntaxError(2,"Unmatched { opened line $lineno[0]",-1);
        $code = substr($$input,$from,pos($$input)-$from-1);
        $lineno[1]+= $code=~tr/\n//;
        return('CODE',[ $code, $lineno[0] ]);
    };

    if($lexlevel == 0) {# In head section
            $$input=~/\G%(left|right|nonassoc)/gc
        and return('ASSOC',[ uc($1), $lineno[0] ]);
            $$input=~/\G%(start)/gc
        and return('START',[ undef, $lineno[0] ]);
            $$input=~/\G%(expect)/gc
        and return('EXPECT',[ undef, $lineno[0] ]);
            $$input=~/\G%\{/gc
        and do {
            my($code);

                $$input=~/\G(.*?)%}/sgc
            or  _SyntaxError(2,"Unmatched %{ opened line $lineno[0]",-1);

            $code=$1;
            $lineno[1]+= $code=~tr/\n//;
            return('HEADCODE',[ $code, $lineno[0] ]);
        };
            $$input=~/\G%(token)/gc
        and return('TOKEN',[ undef, $lineno[0] ]);
            $$input=~/\G%(type)/gc
        and return('TYPE',[ undef, $lineno[0] ]);
            $$input=~/\G%(union)/gc
        and return('UNION',[ undef, $lineno[0] ]);
            $$input=~/\G([0-9]+)/gc
        and return('NUMBER',[ $1, $lineno[0] ]);

    }
    else {# In rule section
            $$input=~/\G%(prec)/gc
        and return('PREC',[ undef, $lineno[0] ]);
    }

    #Always return something
        $$input=~/\G(.)/sg
    or  die "Parse::Yapp::Grammar::Parse: Match (.) failed: report as a BUG";

        $1 eq "\n"
    and ++$lineno[1];

    ( $1 ,[ $1, $lineno[0] ]);

}

sub _SyntaxError {
    my($level,$message,$lineno)=@_;

    $message= "*".
              [ 'Warning', 'Error', 'Fatal' ]->[$level].
              "* $message, at ".
              ($lineno < 0 ? "eof" : "line $lineno").
              ".\n";

        $level > 1
    and die $message;

    warn $message;

        $level > 0
    and ++$nberr;

        $nberr == 20 
    and die "*Fatal* Too many errors detected.\n"
}

sub _AddRules {
    my($lhs,$lineno)=@{$_[0]};
    my($rhss)=$_[1];

        ref($$nterm{$lhs})
    and do {
        _SyntaxError(1,"Non-terminal $lhs redefined: ".
                       "Previously declared line $$syms{$lhs}",$lineno);
        return;
    };

        ref($$term{$lhs})
    and do {
        my($where) = exists($$token{$lhs}) ? $$token{$lhs} : $$syms{$lhs};
        _SyntaxError(1,"Non-terminal $lhs previously ".
                       "declared as token line $where",$lineno);
        return;
    };

        ref($$nterm{$lhs})      #declared through %type
    or  do {
            $$syms{$lhs}=$lineno;   #Say it's declared here
            delete($$term{$lhs});   #No more a terminal
    };
    $$nterm{$lhs}=[];       #It's a non-terminal now

    my($epsrules)=0;        #To issue a warning if more than one epsilon rule

    for my $rhs (@$rhss) {
        my($tmprule)=[ $lhs, [ ], splice(@$rhs,-2) ]; #Init rule

            @$rhs
        or  do {
            ++$$nullable{$lhs};
            ++$epsrules;
        };

        for (0..$#$rhs) {
            my($what,$value)=@{$$rhs[$_]};

                $what eq 'CODE'
            and do {
                my($name)='@'.++$labelno."-$_";
                push(@$rules,[ $name, [], undef, $value ]);
                push(@{$$tmprule[1]},$name);
                next;
            };
            push(@{$$tmprule[1]},$$value[0]);
        }
        push(@$rules,$tmprule);
        push(@{$$nterm{$lhs}},$#$rules);
    }

        $epsrules > 1
    and _SyntaxError(0,"More than one empty rule for symbol $lhs",$lineno);
}

sub Parse {
    my($self)=shift;

        @_ > 0
    or  croak("No input grammar\n");

    my($parsed)={};

    $input=\$_[0];

    $lexlevel=0;
    @lineno=(1,1);
    $nberr=0;
    $prec=0;
    $labelno=0;

    $head=();
    $tail="";

    $syms={};
    $token={};
    $term={};
    $nterm={};
    $rules=[ undef ];   #reserve slot 0 for start rule
    $precterm={};

    $start="";
    $nullable={};
    $expect=0;

    pos($$input)=0;


    $self->YYParse(yylex => \&_Lexer, yyerror => \&_Error);

        $nberr
    and _SyntaxError(2,"Errors detected: No output",-1);

    @$parsed{ 'HEAD', 'TAIL', 'RULES', 'NTERM', 'TERM',
              'NULL', 'PREC', 'SYMS',  'START', 'EXPECT' }
    =       (  $head,  $tail,  $rules,  $nterm,  $term,
               $nullable, $precterm, $syms, $start, $expect);

    undef($input);
    undef($lexlevel);
    undef(@lineno);
    undef($nberr);
    undef($prec);
    undef($labelno);

    undef($head);
    undef($tail);

    undef($syms);
    undef($token);
    undef($term);
    undef($nterm);
    undef($rules);
    undef($precterm);

    undef($start);
    undef($nullable);
    undef($expect);

    $parsed
}


1;
